### ISSUE TITLE
Does not build on GNU Hurd: lib/fts/fts.c:135:48: error: 'MAXPATHLEN' undeclared

### ISSUE THREAD

**Issue by xtaran** [sahib/rmlint#207]
[rmlint fails to build from source on Debian GNU/Hurd](https://buildd.debian.org/status/fetch.php?pkg=rmlint&arch=hurd-i386&ver=2.4.6-1&stamp=1487201702&raw=0) due to using `MAXPATHLEN` (which does not exist on GNU Hurd) unconditionally:

```
gcc -o lib/fts/fts.o -c -Wall -W -Wextra -Winit-self -Wstrict-aliasing -Wmissing-include-dirs -Wuninitialized -Wstrict-prototypes -g -O2 -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -specs=/usr/share/dpkg/pie-compile.specs -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -std=c11 -pipe -D_GNU_SOURCE -fPIC -Os -pthread -I/usr/include/blkid -I/usr/include/uuid -I/usr/include/json-glib-1.0 -I/usr/include/gio-unix-2.0 -I/usr/include/glib-2.0 -I/usr/lib/i386-gnu/glib-2.0/include lib/fts/fts.c
gcc: note: pie specs /usr/share/dpkg/pie-compile.specs ignored when pie is not enabled
In file included from lib/fts/fts.c:40:0:
lib/fts/fts.c: In function 'fts_open':
lib/fts/fts.c:135:48: error: 'MAXPATHLEN' undeclared (first use in this function)
     if(fts_palloc(sp, MAX(fts_maxarglen(argv), MAXPATHLEN)))
                                                ^
lib/fts/fts.c:135:48: note: each undeclared identifier is reported only once for each function it appears in
scons: *** [lib/fts/fts.o] Error 1
scons: building terminated because of errors.
```

This seems to be a widespread and well-known issue, as it's [mentioned  in the GNU Hurd Porting Guidelines](https://www.gnu.org/software/hurd/hurd/porting/guidelines.html#PATH_MAX_tt_MAX_PATH_tt_MAXPATHL).

If someone provides a patch, I can test on a Debian GNU/Hurd porterbox if it fixes the issue.

**Comment by sahib** [comment]
Okay, I see. I wonder how many GNU/Hurd users of ``rmlint`` are out there. :smile:
Will fix that on the weekend, should be easily doable.

**Comment by SeeSpotRun** [comment]
Shouldn't need anything more than https://github.com/SeeSpotRun/rmlint/commit/d6b3995e9ae5549d780ea03ba998fe321eedead0

**Comment by xtaran** [comment]
@SeeSpotRun: `PATH_MAX` isn't available on GNU Hurd either. Reason is that GNU Hurd does not have a limit for path lengths, etc. and hence doesn't declare any related variable.

**Comment by SeeSpotRun** [comment]
I was thinking it would get picked up by https://github.com/sahib/rmlint/blob/master/lib/config.h.in#L50-L56 but on closer inspection maybe not...

https://github.com/SeeSpotRun/rmlint/commit/5321eb08b00a0b41fea8d361e7a1789e827a8ee7 should cover it in any case.  If any actual paths are longer than 4096 then worst case they get missed.

**Comment by xtaran** [comment]
On Thu, Feb 16, 2017 at 06:03:51PM -0800, Daniel T wrote:
> I was thinking it would get picked up by
> https://github.com/sahib/rmlint/blob/master/lib/config.h.in#L50-L56
> but on closer inspection maybe not...

Ah, ok. Missed that rmlint defines its own PATH_MAX if not present.

> https://github.com/SeeSpotRun/rmlint/commit/5321eb08b00a0b41fea8d361e7a1789e827a8ee7
> should cover it in any case.  If any actual paths are longer than
> 4096 then worst case they get missed.

Yeah, but I usually try to avoid such quick and dirty hacks. I've done
them in the past, too, before I realized what's the reason for the
non-existence of these variables on Hurd.

Please have a look at the porting guide I linked to earlier. There is
a description on how to get a clean solution instead of setting
arbitrary limits which may restrict the tool without need. The general
idea is: Don't use static arrays for paths, etc. Use dynamically
allocated ones.

Unfortunately I don't know the fts_* stuff rmlint uses. So it's unclear
to me where to actually apply the according changes.

**Comment by SeeSpotRun** [comment]
Unfortunately there about 8 other modules in rmlint that use PATH_MAX size buffers so that's going to take a little longer to fix.

btw the fts_* stuff is currently an unmodified copy pinched from musl (https://github.com/pullmoll/musl-fts).

**Comment by sahib** [comment]
I agree that variable length pathes would be nice. However I would take the PATH_MAX route for now, since changing it in all places (there are at least 23 of them, each needs to be checked) is not a trivial task. Apart from that, the gain wouldn't be that great beside unlimited path lengths (*) - ``rmlint`` would likely get slower due to many heap allocations and from a quick glance some places need far more code to work nicely.

I will sleep over that, but I think it's unlikely to be implemented.

(*) This could even be seen as security issue... What if some malicious party decides to create a 1GB path?

**Comment by xtaran** [comment]
Hi Chris,

On Fri, Feb 17, 2017 at 02:00:01PM -0800, Chris Pahl wrote:
> I agree that variable length pathes would be nice. However I would
> take the PATH_MAX route for now, since changing it in all places
> (there are at least 23 of them, each needs to be checked) is not a
> trivial task. Apart from that, the gain wouldn't be that great
> beside unlimited path lengths (*)

I see, thanks for the reasoning behind your thoughts on this.

> - ``rmlint`` would likely get slower due to many heap allocations
> and from a quick glance some places need far more code to work
> nicely.

I didn't think about this.

> I will sleep over that, but I think it's unlikely to be implemented.

Understandable.

> (*) This could even be seen as security issue... What if some
> malicious party decides to create a 1GB path?

Probably not to underestimate, but I don't consider being able to chew
on too big data to be a security issue. And I think rmlint is
primarily run on more or less trusted data (the own home directory).

**Comment by sahib** [comment]
I took the PATH_MAX way as of c81f2fc.
See two comments above for the reasoning.

> Probably not to underestimate, but I don't consider being able to chew
> on too big data to be a security issue. And I think rmlint is
> primarily run on more or less trusted data (the own home directory).

That's most likely true, but security always tends to be in places where you least expect it. :smile:
I see that ``rmlint`` is probably not the attack vector, but I was stating that a bit more generally.
Still, for something that is supposed to be a "rock solid" tool (notice I said *supposed*) it should at least
have a (high) hard limit on how big paths may be. This would make a *correct* implementation even harder.

I'll close this issue now, feel free to re-open if the build problems persist on HURD.

**Comment by xtaran** [comment]
JFTR: The (updated) Debian package built fine on GNU Hurd with the hardcoded MAXPATHLEN value: https://buildd.debian.org/status/package.php?p=rmlint

**Comment by sahib** [comment]
@xtaran: Cool, thanks for checking! :+1:
